### PR TITLE
Define ML inputs and outputs methods shared by multiple algorithms

### PR DESCRIPTION
**Pull request by amatissart**
This is a first step before introducing the new algorithm "Mehestan". 

* `MlInput` defines a common interface used to fetch data that can be processed by any algorithm. Two sources are implemented: the database itself and the public dataset .csv file for testing purposes.
* "ml/outputs.py" includes functions to save the individual and entity scores into the database. This is mostly unchanged from "ml_train.py"
* The algorithm used in a poll can be configured in a new field `algorithm`.

### CODE REVIEW

File: backend/ml/inputs.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        Returns:
+        - comparisons_df: DataFrame with columns
+            * `user_id`: int
+            * `entity_a`: int or str
+            * `entity_b`: int or str
+            * `criteria`: str
+            * `score`: float
+            * `weight`: float
+        """
+        raise NotImplementedError
```
Comment (Vikka) [inline backend/ml/inputs.py:22]: Why NotImplementedError ?

Comment (amatissart) [reply backend/ml/inputs.py:22]: Yes. I will define `MlInput` as abstract class to make it more explicit.

File: backend/ml/outputs.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        )
+    else:
+        scores_list = list(contributor_scores)
+
+    rating_ids = {
+        (contributor_id, video_id): rating_id
+        for rating_id, contributor_id, video_id in ContributorRating.objects.filter(
+            poll=poll
+        ).values_list("id", "user_id", "entity_id")
+    }
```
Comment (Vikka) [inline backend/ml/outputs.py:70]:
```suggestion
    rating_ids = {
        (contributor_id, video_id): rating_id 
        for rating_id, contributor_id, video_id in ContributorRating.objects
            .filter(poll=poll).values_list("id", "user_id", "entity_id")
    }
```

File: backend/ml/outputs.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    ratings_to_create = set(
+        (contributor_id, video_id)
+        for contributor_id, video_id, _, _, _ in scores_list
+        if (contributor_id, video_id) not in rating_ids
+    )
```
Comment (Vikka) [inline backend/ml/outputs.py:75]: No need to cast with generator expresion, just use set comprehension.
Also, use a value eater with `*_`, unless you deliberatly want to cause a crash if the number of value is not right.
```suggestion
    ratings_to_create = {(contributor_id, video_id)
                         for contributor_id, video_id, *_ in scores_list
                         if (contributor_id, video_id) not in rating_ids}
```

File: backend/ml/outputs.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    created_ratings = ContributorRating.objects.bulk_create(
+        ContributorRating(
+            poll_id=poll.pk,
+            entity_id=video_id,
+            user_id=contributor_id,
+        )
+        for contributor_id, video_id in ratings_to_create
+    )
```
Comment (Vikka) [inline backend/ml/outputs.py:83]:
```suggestion
        ) for contributor_id, video_id in ratings_to_create)
```

File: backend/ml/outputs.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    rating_ids.update(
+        {(rating.user_id, rating.entity_id): rating.id for rating in created_ratings}
+    )
```
Comment (Vikka) [inline backend/ml/outputs.py:86]:
```suggestion
    rating_ids.update({(rating.user_id, rating.entity_id): rating.id 
                       for rating in created_ratings})
```

File: backend/ml/outputs.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+    scores_to_delete = ContributorRatingCriteriaScore.objects.filter(
+        contributor_rating__poll=poll
+    )
```
Comment (Vikka) [inline backend/ml/outputs.py:90]:
```suggestion
    scores_to_delete = ContributorRatingCriteriaScore.objects \
        .filter(contributor_rating__poll=poll)
```

File: backend/ml/outputs.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    if trusted_filter is not None:
+        if trusted_filter is True:
+            scores_to_delete = scores_to_delete.filter(
+                contributor_rating__user__in=User.trusted_users()
+            )
```
Comment (Vikka) [inline backend/ml/outputs.py:95]:
```suggestion
            scores_to_delete = scores_to_delete \
                .filter(contributor_rating__user__in=User.trusted_users())
```

File: backend/ml/outputs.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    )
+    rating_ids.update(
+        {(rating.user_id, rating.entity_id): rating.id for rating in created_ratings}
+    )
+
+    scores_to_delete = ContributorRatingCriteriaScore.objects.filter(
+        contributor_rating__poll=poll
+    )
+    if trusted_filter is not None:
+        if trusted_filter is True:
```
Comment (Vikka) [inline backend/ml/outputs.py:92]: Never do `is True` or `is False`.
```suggestion
        if trusted_filter:
```

File: backend/ml/outputs.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            scores_to_delete = scores_to_delete.filter(
+                contributor_rating__user__in=User.trusted_users()
+            )
+        elif trusted_filter is False:
+            scores_to_delete = scores_to_delete.exclude(
+                contributor_rating__user__in=User.trusted_users()
+            )
```
Comment (Vikka) [inline backend/ml/outputs.py:99]:
```suggestion
        else:
            scores_to_delete = scores_to_delete \
                .exclude(contributor_rating__user__in=User.trusted_users())
```

File: backend/ml/outputs.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        scores_to_delete.delete()
+        ContributorRatingCriteriaScore.objects.bulk_create(
+            ContributorRatingCriteriaScore(
+                contributor_rating_id=rating_ids[(contributor_id, video_id)],
+                criteria=criteria,
+                score=score,
+                uncertainty=uncertainty,
+            )
+            for contributor_id, video_id, criteria, score, uncertainty in scores_list
+        )
```
Comment (Vikka) [inline backend/ml/outputs.py:113]: 88 c, Black compliant

```suggestion
            ) for contributor_id, video_id, criteria, score, uncertainty in scores_list)
```

File: backend/tournesol/migrations/0023_poll_remove_comparisonsliderchanges_user_and_more.py [hunk 11]
```diff
@@ -6,9 +6,18 @@
...
 import tournesol.models.poll
+from tournesol.models.poll import DEFAULT_POLL_NAME, VideoEntity
 import uuid
 
 
+def create_default_poll(apps, schema_editor):
+    Poll = apps.get_model("tournesol", "Poll")
+    Poll.objects.get_or_create(
+        name=DEFAULT_POLL_NAME,
+        defaults={"entity_type": VideoEntity.name}
```
Comment (Vikka) [inline backend/tournesol/migrations/0023_poll_remove_comparisonsliderchanges_user_and_more.py:17]:
```suggestion
        defaults={"entity_type": VideoEntity.name},
```

**Review by Vikka — CHANGES_REQUESTED** [review]
It took me some time reviewing this, please consider each comment, remember, we write code once, we read it often.

File: backend/ml/inputs.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        Returns:
+        - comparisons_df: DataFrame with columns
+            * `user_id`: int
+            * `entity_a`: int or str
+            * `entity_b`: int or str
+            * `criteria`: str
+            * `score`: float
+            * `weight`: float
+        """
+        pass
```
Comment (Vikka) [inline backend/ml/inputs.py:25]: If the abstract method as a docstring, don't bother with instructions in it.

```suggestion
```

File: backend/ml/inputs.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+
+        Returns:
+        - ratings_df: DataFrame with columns
+            * `user_id`: int
+            * `entity_id`: int or str
+            * `is_public`: bool
+            * `is_trusted`: bool
+            * `is_supertrusted`: bool
+        """
+        pass
```
Comment (Vikka) [inline backend/ml/inputs.py:39]: See previous comment

```suggestion
```

**Review by Vikka — CHANGES_REQUESTED** [review]

**Comment by Vikka** [comment]
@amatissart Why did you force pushed anything ? 

Be careful, with published PR, people might have a local version of your code, and created local commit based on yours. By rebasing and force push, you _migth_ invalidate previous commit and create a huge burden for other contributors.

**Comment by GresilleSiffle** [comment]
> Be careful, with published PR, people might have a local version of your code, and created local commit based on yours. By rebasing and force push, you migth invalidate previous commit and create a huge burden for other contributors.

It's a good point, here is why we chose to be able to force push.

We consider feature branches as rather personal branches, and we don't encourage contributors to work on other contributors branches without syncing. The main benefit is being free to rebase ( we do this a lot ), and force push. Not being able to rebase  and force push because someone else "might" be working on it creates more issues than it solves, the main one is not being able to run the current code rebased on the head of main.

As long as the core team + the contributors allowed to push on the repository are aware of this, everything should be fine.

**Comment by Vikka** [comment]
@GresilleSiffle Thanks ! Good to know !

**Review by GresilleSiffle — APPROVED** [review]

File: backend/ml/inputs.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+    def get_ratings_properties(self):
+        user_entities_pairs = pd.Series(
+            iter(
+                set(self.public_dataset.groupby(["user_id", "entity_a"]).indices.keys())
+                | set(
+                    self.public_dataset.groupby(["user_id", "entity_b"]).indices.keys()
+                )
+            )
+        )
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py:69]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/inputs.py [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+
+
+class MlInputFromDb(MlInput):
+    def __init__(self, poll_name):
+        self.poll_name = poll_name
+
+    def get_comparisons(self, trusted_only=False, criteria=None) -> pd.DataFrame:
+        scores_queryset = ComparisonCriteriaScore.objects.filter(
+            comparison__poll__name=self.poll_name
+        )
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py:84]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/inputs.py [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        if criteria is not None:
+            scores_queryset = scores_queryset.filter(criteria=criteria)
+
+        if trusted_only:
+            scores_queryset = scores_queryset.filter(
+                comparison__user__in=User.trusted_users()
+            )
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py:91]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/inputs.py [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+                    "score",
+                    "weight",
+                ]
+            )
+
+    def get_ratings_properties(self):
+        values = (
+            ContributorRating.objects.filter(
+                poll__name=self.poll_name,
+            )
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py:123]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/outputs.py (hunk 2 above)
Comment (GresilleSiffle) [reply backend/ml/outputs.py:75]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/outputs.py (hunk 3 above)
Comment (GresilleSiffle) [reply backend/ml/outputs.py:80]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/outputs.py (hunk 5 above)
Comment (GresilleSiffle) [reply backend/ml/outputs.py:91]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/outputs.py (hunk 7 above)
Comment (GresilleSiffle) [reply backend/ml/outputs.py:100]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/outputs.py (hunk 9 above)
Comment (GresilleSiffle) [reply backend/ml/outputs.py]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/inputs.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+class MlInputFromPublicDataset(MlInput):
+    def __init__(self, csv_file):
+        self.public_dataset = pd.read_csv(csv_file)
+        self.public_dataset.rename(
+            {"video_a": "entity_a", "video_b": "entity_b"}, axis=1, inplace=True
+        )
+        self.public_dataset["user_id"], self.user_indices = self.public_dataset[
+            "public_username"
+        ].factorize()
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py:53]: Saw with Adrien.

Thanks again for the feedback.

We're going to keep this as is for now, as this suggested change will break the consistency between this file and the rest of the module / project.

We will consider updating those line breaks after the deliberation about the module `black`.

File: backend/ml/inputs.py [hunk 19]
```diff
@@ -0,0 +1,142 @@
+import pandas as pd
+from django.db.models import Case, F, When
+
+from core.models import User
+from tournesol.models import ComparisonCriteriaScore
+from tournesol.models.ratings import ContributorRating
+
+
+class MlInput:
+    def get_comparisons(self, trusted_only=False, criteria=None) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/712/commits/a8cdc506f3c3d503f3b790e176b35175a8be37db.

File: backend/ml/inputs.py [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+        df = pd.DataFrame([*user_entities_pairs], columns=["user_id", "entity_id"])
+        df["is_public"] = True
+        top_users = df.value_counts("user_id").index[:6]
+        df["is_trusted"] = df["is_supertrusted"] = df["user_id"].isin(top_users)
+        return df
+
+
+class MlInputFromDb(MlInput):
+    def __init__(self, poll_name):
+        self.poll_name = poll_name
```
[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py]: Fixed by https://github.com/tournesol-app/tournesol/pull/712/commits/a8cdc506f3c3d503f3b790e176b35175a8be37db.

File: backend/ml/inputs.py [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+                comparison__user__in=User.trusted_users()
+            )
+
+        values = scores_queryset.values(
+            "score",
+            "criteria",
+            "weight",
+            entity_a=F("comparison__entity_1_id"),
+            entity_b=F("comparison__entity_2_id"),
+            user_id=F("comparison__user_id"),
```
[the remark this answers is not in the document]
Comment (amatissart) [reply backend/ml/inputs.py:99]: `F` is used to build a SQL expression that represents a model field:
https://docs.djangoproject.com/en/4.0/ref/models/expressions/#f-expressions

[the remark this answers is not in the document]
Comment (GresilleSiffle) [reply backend/ml/inputs.py:99]: `F()` is a class parts of the Django query expressions. Instances of `F` represent value of a  model field.

See: https://docs.djangoproject.com/en/4.0/ref/models/expressions/#f-expressions

Also see the imports statements of this file.

**Pull request merged**